source/common/http/header_map_impl.cc

```diff
@@ -443,57 +434,79 @@ void HeaderMapImpl::addCopy(const LowerCaseString& key, uint64_t value) {
     return;
   }
   HeaderString new_key;
-  new_key.setCopy(key.get().c_str(), key.get().size());
+  new_key.setCopy(key.get());
   HeaderString new_value;
   new_value.setInteger(value);
   insertByKey(std::move(new_key), std::move(new_value));
   ASSERT(new_key.empty());   // NOLINT(bugprone-use-after-move)
   ASSERT(new_value.empty()); // NOLINT(bugprone-use-after-move)
 }
 
-void HeaderMapImpl::addCopy(const LowerCaseString& key, const std::string& value) {
+void HeaderMapImpl::addCopy(const LowerCaseString& key, absl::string_view value) {
   auto* entry = getExistingInline(key.get());
   if (entry != nullptr) {
     const uint64_t added_size = appendToHeader(entry->value(), value);
     addSize(added_size);
     return;
   }
   HeaderString new_key;
-  new_key.setCopy(key.get().c_str(), key.get().size());
+  new_key.setCopy(key.get());
   HeaderString new_value;
-  new_value.setCopy(value.c_str(), value.size());
+  new_value.setCopy(value);
   insertByKey(std::move(new_key), std::move(new_value));
   ASSERT(new_key.empty());   // NOLINT(bugprone-use-after-move)
   ASSERT(new_value.empty()); // NOLINT(bugprone-use-after-move)
 }
 
-void HeaderMapImpl::setReference(const LowerCaseString& key, const std::string& value) {
+void HeaderMapImpl::append(const LowerCaseString& key, absl::string_view value) {
+  bool found = false;
+  for (HeaderEntryImpl& header : headers_) {
```

Comment:
Do we have to iterate here? I worry about O(n^2) potential.

Reply:
I brought this up in the PR description / this was the header filter the dilemma. Suppose you have a HeaderMap with
`a: foo; a: bar`
And you configure authz filter to append `baz` to `a`, would it suffice to make the HeaderMap
`a: foo, baz; a: bar?

For reference, `setCopy` is used in lua wrappers and ext authz and `append` is used in ext authz. @mattklein123 might have some context about the lua wrappers (or maybe not, although you might have written it i think it was in response to a user issue).

Reply:
I don't think there is any N^2 issue here, it's a linear scan through the headers which we also do in the non O(1) get path, etc. I know there has been discussion here around adding a parallel hash set for headers but as we have already discussed the performance implication of that is unclear. You could put in a TODO to investigate if you want here and in other relevant spots. Up to you.

With all of that said, In the case that you identify above, technically I think we could collapse the headers and concatenate them per the RFC. Might be worth a TODO/comment for that also?

Reply:
Yeah, we do this elsewhere. Can you remind me why this isn't an O(n^2) issue if we are using these methods from some other code that is also iterating over the header map?

Reply:
I suggest it's OK if this is an O(N) operation for now, as long as it's documented, and if someone is making this call in a loop over headers we should have a higher level discussion about how to make the aggregated operation O(N).

Reply:
I think we have a problematic structural issue here though; header maps are consumed by arbitrary filters. We're providing a dangerous operation to anyone who writes a filter, even filters we don't see. We should at a minimum flag this with documentation, big warnings on the interface and also open some issues and add TODOs. We know O(n^2) issues can be real vulnerabilities on the data path, so let's not tempt fate.

Reply:
Yes it's n^2 if the caller is looping. I don't think we should be trying to fix this issue in this PR, so I would open an issue to track. IMO clear comments are fine for now. 

We already have some mitigations in terms of max header map size. 

In the future we could consider max sizes for header keys, potentially adding an internal hash set, utilities that build a hash set and then perform operations, etc. But none of these things are trivial and all have substantial tradeoffs.